### PR TITLE
Graceful connection shutdown in CLI

### PR DESCRIPTION
**Pull request by anton-ryzhov**
Simplify and improve the CLI tool —  close the connection gracefully, handle disconnects.

`Ctrl-C`, `Ctrl-D` or `exit()` ways of exit are handled.

Remote end disconnection causes CLI to exit immediately

### CODE REVIEW

File: src/manhole/__init__.py [hunk 1]
```diff
@@ -363,6 +363,8 @@ def handle_repl(locals):
         namespace.update(locals)
     try:
         ManholeConsole(namespace).interact()
+    except SystemExit:
```
Comment (anton-ryzhov) [inline src/manhole/__init__.py:366]: `exit()` called in console was logged as unexpected error

File: src/manhole/cli.py [hunk 2]
```diff
@@ -69,41 +68,26 @@ def parse_signal(value):
...
-                data = os.read(self.read_fd, 1024)
-                self.sock.sendall(data)
-            else:
-                raise RuntimeError("Unknown FD %s" % fd)
+            except socket.timeout:
+                pass
+
+        if not self.is_closing.is_set():
+            # Break waiting for input()
+            os.kill(os.getpid(), signal.SIGINT)
```
Comment (anton-ryzhov) [inline src/manhole/cli.py:90]: If main thread is still in `input()` — break the waiting loog

File: src/manhole/cli.py [hunk 3]
```diff
@@ -138,21 +122,19 @@ def main():
...
-    except KeyboardInterrupt:
+            data = input()
+            data += '\n'
+            sock.sendall(data.encode('utf8'))
+    except (EOFError, KeyboardInterrupt):
         pass
     finally:
-        thread.should_run = False
+        is_closing.set()
+        sock.shutdown(socket.SHUT_WR)
```
Comment (anton-ryzhov) [inline src/manhole/cli.py:138]: Send EOF to remote end to let it terminate the connection (send TCP FIN)

File: src/manhole/cli.py [hunk 4]
```diff
@@ -138,21 +122,19 @@ def main():
...
         thread.join()
```
Comment (anton-ryzhov) [inline src/manhole/cli.py:139]: Wait for remote end to close the connection

**Comment by anton-ryzhov** [comment]
JFYI This is the last PR in my backlog

**Comment by ionelmc** [comment]
Did the timeout option disappear completely? I think the timeout stuff wasn't tested unfortunately, so tests passed.

**Comment by anton-ryzhov** [comment]
It is [still used](https://github.com/anton-ryzhov/python-manhole/blob/facb2c16cda98989414e235c0a5cfe2144cfc7cd/src/manhole/cli.py#L112).

But I'm not sure how useful it is now. On one hand that makes [this loop](https://github.com/ionelmc/python-manhole/pull/63/files#diff-aa51e4fce1214f64e544342d483686dc752e452781efda52daa43f2650afc3b1R77) to do noop cycles instead of just waiting until the data comes. On the other hand it may[be] stuck there forever (?) in some corner cases…

I just kept it there because I wasn't 100% sure

**Comment by ionelmc** [comment]
I've tried it a bit, seems to work as expected.

**Comment by anton-ryzhov** [comment]
For some reason last release is not marked as latest
image

**Pull request merged**